torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,1736 @@
+            self.reshard_after_forward = False
+
+            # Due to the use of streams, we need to make sure the previous
+            # ``optim.step()`` is done before we all-gather parameters.
+            self._wait_for_previous_optim_step()
+
+    @torch.no_grad()
+    def _init_param_attributes(self, p: Parameter) -> None:
+        """
+        We manage several attributes on each Parameter instance. The first two
+        are set by :func:`_shard_parameters_`:
+            ``_is_sharded``: ``True`` if the Parameter is sharded or ``False``
+                if the Parameter is intentionally not sharded (in which case we
+                will all-reduce grads for this param).
+            ``_orig_size``: the size of the original Parameter (before sharding)
+        The remaining attributes are set here:
+            ``_fp32_shard``: a single shard of the parameters in full precision
+                (typically FP32, but this is dependent on the dtype of the model
+                as it's passed in by the user). This can be on CPU or GPU
+                depending on the value of *``cpu_offload``*.
+            ``_fp16_shard``: if *``mixed_precision``* is ``True``, this will be
+                a single shard of the parameters in FP16, used for all-gather.
+            ``_full_param_padded``: the full weight (padded to be evenly
+                divisible by ``world_size``), used for computation in the
+                forward and backward pass. This will be resized in place and
+                only materialized (via all-gather) as needed.
+        """
+        assert hasattr(p, "_is_sharded") and hasattr(p, "_orig_size")
+        if hasattr(p, "_fp32_shard"):
+            return
+
+        # A single shard of the parameters in full precision.
+        p._fp32_shard = p.data  # type: ignore[attr-defined]
+
+        if self.mixed_precision:
+            assert p._fp32_shard.dtype == torch.float32  # type: ignore[attr-defined]
+
+            if self.move_params_to_cpu:
+                assert p._fp32_shard.device == torch.device("cpu")  # type: ignore[attr-defined]
+                # If we plan to keep the FP32 parameters on CPU, then pinning
+                # memory allows us to later use non-blocking transfers when moving
+                # the FP32 param shard to compute_device.
+                p._fp32_shard = p._fp32_shard.pin_memory()  # type: ignore[attr-defined]
+                p.data = p._fp32_shard  # type: ignore[attr-defined]
+
+            # In mixed precision mode, we maintain a reduced precision
+            # (typically FP16) parameter shard on compute_device for performing
+            # the computation in the forward/backward pass. We resize the
+            # storage to size 0 at init (here) and re-materialize (by copying
+            # from _fp32_shard) as needed.
+            p._fp16_shard = torch.zeros_like(  # type: ignore[attr-defined]
+                p._fp32_shard,  # type: ignore[attr-defined]
```

Comment:
curious, why do we need to keep the `_fp32_shard` here? And how does this work with AMP's mixed precision wrapper?

Reply:
my understanding is: _fp32_shard is the original sharded parameters that will be updated in optimizer, when forward and backward are done in fsdp, parameter.data needs to point back to _fp32_shard for updating.  

if mixed precision=True, in default it will convert fp32 to fp16 before forward all gather, since then all the forward/backward computation and communication are done using fp16, after grad is reduce_scattered, the local fp16 shard of grad will be converted back to fp32 for optimizer updating.  In this case, AMP's mixed precision will be a no-op (my guess), as operators' input is always fp16, no conversion is needed from AMP.  

if mixed_precision = false, similar to ddp, AMP's mixed precision will do the auto cast